x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java

```diff
@@ -0,0 +1,595 @@
+            a = MinimizationOperations.minimize(a, Integer.MAX_VALUE);
+
+            // TODO make this a param and scale up based on num shards like we do with terms aggs?
+            int shard_size = request.size();
+
+            CompiledAutomaton automaton = new CompiledAutomaton(a);
+            TermsEnum te = automaton.getTermsEnum(terms);
+
+            // All the above prep might take a while - do a timer check now before we continue further.
+            if (System.currentTimeMillis() > scheduledEnd) {
+                return new ShardTermEnumResponse(request.shardId(), termsList, error, true);
+            }
+
+            int numTermsBetweenClockChecks = 100;
+            int termCount = 0;
+            if (request.sortByPopularity()) {
+                // Collect most popular matches
+                TermCountPriorityQueue pq = new TermCountPriorityQueue(shard_size);
+                TermCount spare = null;
+                while (te.next() != null) {
+                    termCount++;
+                    if (termCount > numTermsBetweenClockChecks) {
+                        if (System.currentTimeMillis() > scheduledEnd) {
+                            // Gather what we have collected so far
+                            while (pq.size() > 0) {
+                                termsList.add(pq.pop());
+                            }
+                            return new ShardTermEnumResponse(request.shardId(), termsList, error, true);
+                        }
+                        termCount = 0;
+                    }
+                    int df = te.docFreq();
+                    BytesRef bytes = te.term();
+
+                    if (spare == null) {
+                        spare = new TermCount(bytes.utf8ToString(), df);
+                    } else {
+                        spare.setTerm(bytes.utf8ToString());
+                        spare.setDocCount(df);
+                    }
+                    spare = pq.insertWithOverflow(spare);
+                }
+                while (pq.size() > 0) {
+                    termsList.add(pq.pop());
+                }
+            } else {
+                // Collect in alphabetical order
+                while (te.next() != null) {
+                    termCount++;
+                    if (termCount > numTermsBetweenClockChecks) {
+                        if (System.currentTimeMillis() > scheduledEnd) {
+                            return new ShardTermEnumResponse(request.shardId(), termsList, error, true);
+                        }
+                        termCount = 0;
+                    }
+                    int df = te.docFreq();
+                    BytesRef bytes = te.term();
+                    termsList.add(new TermCount(bytes.utf8ToString(), df));
+                    if (termsList.size() >= shard_size) {
+                        break;
+                    }
+                }
+            }
+
+        } catch (AssertionError e) {
```

Comment:
we should never catch an AssertionError, did you mean `Exception` ?

Reply:
Yep - a copy and paste from TransportValidateQueryAction